src/box/txn_limbo.c

```diff
@@ -490,16 +496,25 @@ txn_limbo_read_rollback(struct txn_limbo *limbo, int64_t lsn)
 void
 txn_limbo_write_promote(struct txn_limbo *limbo, int64_t lsn, uint64_t term)
 {
-	limbo->confirmed_lsn = lsn;
-	limbo->is_in_rollback = true;
+	assert(latch_is_locked(&limbo->promote_latch));
 	/*
 	 * We make sure that promote is only written once everything this
 	 * instance has may be confirmed.
 	 */
 	struct txn_limbo_entry *e = txn_limbo_last_synchro_entry(limbo);
 	assert(e == NULL || e->lsn <= lsn);
 	(void) e;
-	txn_limbo_write_synchro(limbo, IPROTO_RAFT_PROMOTE, lsn, term);
+	struct synchro_request req = {
+		.type = IPROTO_RAFT_PROMOTE,
+		.replica_id = limbo->owner_id,
+		.origin_id = instance_id,
```

Comment:
Unfortunately, we can't keep using `txn_limbo_write_synchro()` in this place. The reason is that I need to prepare a request before it goes to WAL. Which means I need `synchro_request` at hand before calling the write. Are you worried about people filling unnecessary fields like I presumably did here? Maybe it is enough that it is now encapsulated in txn_limbo.c? At least, it is not in box.cc anymore.

The `origin_id` is needed here because I call `txn_limbo_req_commit()` next and it needs the field filled. Uses it for term lookup. It was called before too though.

Reply:
Ok, I see, thanks!